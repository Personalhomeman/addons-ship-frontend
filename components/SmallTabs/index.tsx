import React, { useRef, useLayoutEffect, useState } from 'react';
import {
  Flex,
  Text,
  PlacementManager,
  PlacementReference,
  Icon,
  Placement,
  DropdownMenu,
  DropdownMenuItem
} from '@bitrise/bitkit';
import cx from 'classnames';

import css from './style.scss';

type Props = {
  items: Array<{
    key: string;
    value: string;
  }>;
  selected: string;
  onSelect: (key: string) => void;
};

export default ({ items: originalItems, selected, onSelect }: Props) => {
  const dropdownButtonWidth = 40;
  const container = useRef(null);

  const [isOpen, setOpen] = useState(false);
  const [hasOwerflow, setHasOwerflow] = useState(false);
  const [items, setItems] = useState(
    originalItems.map(item => ({
      ...item,
      ref: useRef(null),
      isVisible: false
    }))
  );

  useLayoutEffect(() => {
    const { right: containerRight } = (container as any).current.getBoundingClientRect();

    setItems(items =>
      items.map(item => {
        const { right } = (item.ref as any).current.getBoundingClientRect();

        return { ...item, isVisible: containerRight - dropdownButtonWidth > right };
      })
    );

    setHasOwerflow(() => items.some(({ isVisible }) => !isVisible));
  }, [originalItems, selected]);

  return (
    <Flex direction="horizontal" className={css.container} innerRef={container} alignChildrenVertical="middle">
      <Flex direction="horizontal" style={{ flex: 1, overflow: 'hidden' }}>
        <Flex direction="horizontal" className={css.list}>
          {items.map(({ key, value, ref, isVisible }) => (
            <Text
              key={key}
              className={cx(css.item, { [css.itemSelected]: key === selected, [css.itemHidden]: !isVisible })}
              weight="medium"
              size="x3"
              color="grape-4"
              onClick={() => onSelect(key)}
            >
              <span ref={ref}>{value}</span>
            </Text>
          ))}
        </Flex>
      </Flex>
      {hasOwerflow && (
        <PlacementManager>
          <PlacementReference>
            {({ ref }) => (
              <div className={css.dropdown} ref={ref} onClick={() => setOpen(isOpen => !isOpen)}>
                <Icon name="ChevronDown" />
              </div>
            )}
          </PlacementReference>
          <Placement onClose={() => setOpen(false)} visible={isOpen} placement="bottom-end">
            {() => (
              <DropdownMenu maxHeight="16rem" width="12rem">
                {items
                  .filter(({ isVisible }) => !isVisible)
                  .map(({ key, value }) => (
                    <DropdownMenuItem
                      key={key}
                      onClick={() => onSelect(key)}
                      className="nope"
                      paddingHorizontal="x0"
                      selected={key === selected}
                    >
                      {value}
                    </DropdownMenuItem>
                  ))}
              </DropdownMenu>
            )}
          </Placement>
        </PlacementManager>
      )}
    </Flex>
  );
};
