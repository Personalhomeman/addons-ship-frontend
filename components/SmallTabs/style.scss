@import '~@/assets/style/colors';
@import '~@/assets/style/dimensions';

.container {
  position: relative;
  margin-top: 2 * $grid;
  margin-bottom: 3 * $grid;
  overflow-x: hidden;
}

.list {
  width: 9999px;
}

.item {
  padding: 2 * $grid;
  border-radius: $grid;
  cursor: pointer;
  transition: background 0.2s, color 0.2s;

  &:hover {
    background: $grape-1;
    color: $grape-3;
  }

  &Selected::after {
    content: '';
    display: block;
    height: 2px;
    width: 100%;
    background: $aqua-3;
  }

  &Hidden {
    opacity: 0;
    pointer-events: none;
  }

  &Marked:before {
    content: '';
    display: inline-block;
    width: $grid;
    height: $grid;
    margin-right: $grid;
    margin-bottom: $grid;
    border-radius: $grid / 2;
    background-color: $aqua-3;
  }
}

.dropdown {
  padding: $grid;
  border-radius: $grid;
  cursor: pointer;
  transition: background 0.2s, color 0.2s;

  &:hover {
    background: $grape-1;
    color: $grape-3;
  }
}

:global(.Placement) {
  margin: 3 * $grid 0;
  transition: none;
}
