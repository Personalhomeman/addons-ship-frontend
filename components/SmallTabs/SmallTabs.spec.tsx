// const mockUseStateSet = jest.fn();

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(input => [input, jest.fn()]),
  useRef: jest.fn(),
  useLayoutEffect: jest.fn(),
  useEffect: jest.fn()
}));

import React, { useRef, useState, useLayoutEffect } from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

// import { asMock } from '@/utils';

import SmallTabs from './';

describe('SmallTabs', () => {
  const items = [
    {
      key: 'item-1',
      value: 'Item 1'
    },
    {
      key: 'item-2',
      value: 'Item 2'
    }
  ];

  let mockGetBoundingClientRect = jest.fn();

  beforeEach(() => {
    (useRef as jest.Mock).mockImplementation(() => ({
      current: { getBoundingClientRect: mockGetBoundingClientRect }
    }));
  });

  afterEach(() => {
    mockGetBoundingClientRect.mockRestore();
    (useLayoutEffect as jest.Mock).mockRestore();
  });

  it('renders correctly', () => {
    const tree = shallowToJson(shallow(<SmallTabs items={items} selected={items[0].key} onSelect={jest.fn()} />));

    expect(tree).toMatchSnapshot();
  });

  it('calculates useLayoutEffect contents', () => {
    mockGetBoundingClientRect.mockReturnValueOnce({ right: 0 });
    (useLayoutEffect as jest.Mock).mockImplementationOnce(fn => fn());

    <SmallTabs items={items} selected={items[0].key} onSelect={jest.fn()} />;
  });

  it('shows a dropdown if any menu item is not visible', () => {
    (useState as jest.Mock)
      .mockImplementationOnce(() => [false, jest.fn()]) // isOpen
      .mockImplementationOnce(() => [true, jest.fn()]); // hasOwerflow

    const tree = shallowToJson(shallow(<SmallTabs items={items} selected={items[0].key} onSelect={jest.fn()} />));

    expect(tree).toMatchSnapshot();
  });

  test('onSelect', () => {
    const onSelect = jest.fn();
    const wrap = shallow(<SmallTabs items={items} selected={items[0].key} onSelect={onSelect} />);

    wrap
      .find('.item')
      .last()
      .simulate('click');

    expect(onSelect).toHaveBeenCalledWith('item-2');
  });

  test('dropdown onSelect', () => {
    const onSelect = jest.fn();
    (useState as jest.Mock)
      .mockImplementationOnce(() => [true, jest.fn()]) // isOpen
      .mockImplementationOnce(() => [true, jest.fn()]); // hasOwerflow

    const wrap = mount(<SmallTabs items={items} selected={items[0].key} onSelect={onSelect} />);

    wrap
      .find('DropdownMenuItem')
      .last()
      .simulate('click', {
        nativeEvent: {
          stopImmediatePropagation: jest.fn()
        }
      });

    expect(onSelect).toHaveBeenCalledWith('item-2');
  });

  test.todo(
    'onClose'
    // , () => {
    //   const setOpen = jest.fn();
    //   (useState as jest.Mock).mockImplementationOnce(() => [true, setOpen]); // isOpen

    //   const wrap = shallow(<SmallTabs items={items} selected={items[0].key} onSelect={jest.fn()} />);

    //   wrap.first().simulate('click'); // clickout

    //   expect(setOpen).toHaveBeenCalledWith(false);
    // }
  );
});
