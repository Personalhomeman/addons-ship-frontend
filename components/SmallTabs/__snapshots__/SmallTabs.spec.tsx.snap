// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmallTabs renders correctly 1`] = `
<Flex
  alignChildrenVertical="middle"
  className="container"
  direction="horizontal"
  innerRef={
    Object {
      "current": Object {
        "getBoundingClientRect": [MockFunction],
      },
    }
  }
>
  <Flex
    direction="horizontal"
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <Flex
      className="list"
      direction="horizontal"
    >
      <Text
        className="item itemSelected itemHidden"
        color="grape-4"
        key="item-1"
        onClick={[Function]}
        size="x3"
        weight="medium"
      >
        <span>
          Item 1
        </span>
      </Text>
      <Text
        className="item itemHidden"
        color="grape-4"
        key="item-2"
        onClick={[Function]}
        size="x3"
        weight="medium"
      >
        <span>
          Item 2
        </span>
      </Text>
    </Flex>
  </Flex>
</Flex>
`;

exports[`SmallTabs shows a dropdown if any menu item is not visible 1`] = `
<Flex
  alignChildrenVertical="middle"
  className="container"
  direction="horizontal"
  innerRef={
    Object {
      "current": Object {
        "getBoundingClientRect": [MockFunction],
      },
    }
  }
>
  <Flex
    direction="horizontal"
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <Flex
      className="list"
      direction="horizontal"
    >
      <Text
        className="item itemSelected itemHidden"
        color="grape-4"
        key="item-1"
        onClick={[Function]}
        size="x3"
        weight="medium"
      >
        <span>
          Item 1
        </span>
      </Text>
      <Text
        className="item itemHidden"
        color="grape-4"
        key="item-2"
        onClick={[Function]}
        size="x3"
        weight="medium"
      >
        <span>
          Item 2
        </span>
      </Text>
    </Flex>
  </Flex>
  <PlacementManager>
    <PlacementReference>
      <Component />
    </PlacementReference>
    <Placement
      onClose={[Function]}
      placement="bottom-end"
      visible={false}
    >
      <Component />
    </Placement>
  </PlacementManager>
</Flex>
`;
